processing/src/main/java/io/druid/segment/DoubleDimensionIndexer.java

```diff
@@ -0,0 +1,258 @@
+      }
+
+      @Override
+      public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+      {
+        // nothing to inspect
+      }
+    }
+
+    return new IndexerLongColumnSelector();
+  }
+
+  @Override
+  public FloatColumnSelector makeFloatColumnSelector(
+      IncrementalIndexStorageAdapter.EntryHolder currEntry, IncrementalIndex.DimensionDesc desc
+  )
+  {
+    final int dimIndex = desc.getIndex();
+    class IndexerFloatColumnSelector implements FloatColumnSelector
+    {
+      @Override
+      public float get()
+      {
+        final Object[] dims = currEntry.getKey().getDims();
+
+        if (dimIndex >= dims.length) {
+          return 0.0f;
+        }
+
+        double doubleValue = (Double) dims[dimIndex];
+        return (float) doubleValue;
+      }
+
+      @Override
+      public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+      {
+        // nothing to inspect
+      }
+    }
+
+    return new IndexerFloatColumnSelector();
+  }
+
+  @Override
+  public ObjectColumnSelector makeObjectColumnSelector(
+      DimensionSpec spec, IncrementalIndexStorageAdapter.EntryHolder currEntry, IncrementalIndex.DimensionDesc desc
+  )
+  {
+    final int dimIndex = desc.getIndex();
+    class IndexerObjectColumnSelector implements ObjectColumnSelector
+    {
+      @Override
+      public Class classOfObject()
+      {
+        return Double.class;
+      }
+
+      @Override
+      public Object get()
+      {
+        final Object[] dims = currEntry.getKey().getDims();
+
+        if (dimIndex >= dims.length) {
+          return DimensionHandlerUtils.ZERO_DOUBLE;
+        }
+
+        return dims[dimIndex];
+      }
+    }
+
+    return new IndexerObjectColumnSelector();
+  }
+
+  @Override
+  public DoubleColumnSelector makeDoubleColumnSelector(
+      IncrementalIndexStorageAdapter.EntryHolder currEntry, IncrementalIndex.DimensionDesc desc
+  )
+  {
+    final int dimIndex = desc.getIndex();
+    class IndexerDoubleColumnSelector implements DoubleColumnSelector
+    {
+      @Override
+      public double get()
+      {
+        final Object[] dims = currEntry.getKey().getDims();
+
+        if (dimIndex >= dims.length) {
+          return 0.0;
+        }
+        return (Double) dims[dimIndex];
+      }
+
+      @Override
+      public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+      {
+        // nothing to inspect
+      }
+    }
+
+    return new IndexerDoubleColumnSelector();
+  }
+
+  @Override
+  public int compareUnsortedEncodedKeyComponents(@Nullable Double lhs, @Nullable Double rhs)
+  {
+    // DimensionHandlerUtils.convertObjectToDouble is used to convert null to DoubleZero
+    return Double.compare(DimensionHandlerUtils.convertObjectToDouble(lhs), DimensionHandlerUtils.convertObjectToDouble(rhs));
+  }
+
+  @Override
+  public boolean checkUnsortedEncodedKeyComponentsEqual(@Nullable Double lhs, @Nullable Double rhs)
+  {
+    return Objects.equals(lhs, rhs);
```

Comment:
If we decide to make null totally equivalent to zero, it should probably be nullToZero(lhs).equals(nullToZero(rhs)`. Otherwise `null` won't match zero.